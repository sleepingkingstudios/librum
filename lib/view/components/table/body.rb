# frozen_string_literal: true

module View::Components
  # Component for rendering a table body.
  class Table::Body < ViewComponent::Base
    # @param columns [Array<ColumnDefinition>] the columns used to render the
    #   table.
    # @param data [Array<Hash{String=>Object}>] the table data to render.
    # @param empty_message [String, ViewComponent::Base] the message or
    #   component to display when the table has no data.
    # @param cell_component [ViewComponent::Base] the component to render each
    #   table cell. Defaults to View::Components::Table::Cell.
    # @param row_component [ViewComponent::Base] the component to render each
    #   table row. Defaults to View::Components::Table::Row.
    # @param options [Hash{Symbol=>Object}] additional options to pass to the
    #   rendered table body.
    def initialize( # rubocop:disable Metrics/ParameterLists
      columns:,
      data:,
      cell_component: nil,
      empty_message:  nil,
      row_component:  nil,
      **options
    )
      super()

      @cell_component = cell_component || View::Components::Table::Cell
      @columns        = columns
      @data           = data
      @empty_message  = empty_message
      @row_component  = row_component || View::Components::Table::Row
      @options        = options
    end

    # @return [ViewComponent::Base] the component to render each table cell.
    attr_reader :cell_component

    # @return [Array<ColumnDefinition>] the columns used to render the table.
    attr_reader :columns

    # @return [Array<Hash{String=>Object}>] the table data to render.
    attr_reader :data

    # @return [String, ViewComponent::Base] the message or component to display
    #   when the table has no data.
    attr_reader :empty_message

    # @return [Hash{Symbol=>Object}] additional options to pass to the rendered
    #   table body
    attr_reader :options

    # @return [ViewComponent::Base] the component to render each table row.
    attr_reader :row_component

    private

    def build_empty_message
      View::Components::Table::EmptyMessage.new(
        columns:       columns,
        empty_message: empty_message,
        **options
      )
    end

    def build_row(item:)
      row_component.new(
        cell_component: cell_component,
        columns:        columns,
        item:           item,
        **options
      )
    end

    def render_empty_message
      return render(empty_message) if empty_message.is_a?(ViewComponent::Base)

      render(build_empty_message)
    end

    def render_row(item:)
      render(build_row(item: item))
    end
  end
end
