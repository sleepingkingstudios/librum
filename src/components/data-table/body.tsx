import * as React from 'react';
import { kebabCase } from 'lodash';

import { DataTableRow } from './row';
import type { DataTableColumn } from './types';

const emptyMessageFor = ({
  emptyMessage,
  name,
}: {
  emptyMessage: React.ComponentType | string,
  name: string,
}): JSX.Element | string => {
  if (typeof emptyMessage === 'function') {
    const EmptyMessage = emptyMessage;

    return (<EmptyMessage />);
  }

  if (typeof emptyMessage === 'string') { return emptyMessage; }

  return `There are no matching ${kebabCase(name).replace(/-/g, ' ')}.`;
};

export const DataTableBody = ({
  columns,
  data,
  emptyMessage,
  name,
}: {
  columns: DataTableColumn[],
  data: Record<string, unknown>[],
  emptyMessage?: React.ComponentType | string,
  name: string,
}): JSX.Element => {
  if (data.length === 0) {
    return (
      <tbody>
        <tr>
          <td className="data-table-empty" colSpan={columns.length}>
            { emptyMessageFor({ emptyMessage, name }) }
          </td>
        </tr>
      </tbody>
    );
  }

  return (
    <tbody>
      {
        data.map(
          (rowData: Record<string, unknown>, index: number) => (
            <DataTableRow
              columns={columns}
              data={rowData}
              key={`${name}-${index}`}
            />
          )
        )
      }
    </tbody>
  );
};
