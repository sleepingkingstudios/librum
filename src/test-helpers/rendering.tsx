import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
