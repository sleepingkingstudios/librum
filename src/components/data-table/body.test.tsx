import * as React from 'react';

import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { DataTableBody } from './body';
import type { DataTableColumn } from './types';

type SpaceMission = {
  deltaV: string,
  launchDate?: string,
  missionName?: string,
  missionObjective: string,
  ordinal: string,
  program: string,
};

const BodyWrapper = ({
  children,
}: {
  children: React.ReactNode,
}): JSX.Element => (<table>{ children }</table>);

describe('<DataTableBody />', () => {
  const Actions = (): JSX.Element => (
    <span className="data-table-actions">Launch</span>
  );
  const columns: DataTableColumn[] = [
    {
      name: 'missionName',
      value: ({ data }: { data: SpaceMission }) => (
        `${data.program} ${data.ordinal}`
      ),
    },
    {
      label: 'Objective',
      name: 'missionObjective',
    },
    {
      name: 'deltaV',
      numeric: true,
    },
    {
      align: 'center',
      default: 'TBD',
      name: 'launchDate',
    },
    {
      label: false,
      name: 'actions',
      value: Actions,
    },
  ];
  const name = 'spaceMissions';

  describe('when there are no rows', () => {
    const data: SpaceMission[] = [];
    const expected = 'There are no matching space missions.';

    it('should display the empty message', () => {
      const { getByRole } = render(
        <DataTableBody columns={columns} data={data} name={name} />,
        { wrapper: BodyWrapper },
      );
      const cell = getByRole('cell');

      expect(cell.textContent).toBe(expected);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableBody columns={columns} data={data} name={name} />,
        { wrapper: BodyWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there is one row', () => {
    const data: SpaceMission[] = [
      {
        deltaV: '1,000 km/s',
        missionObjective: '40 Eridani A',
        ordinal: 'VI',
        program: 'Hephaestus',
      },
    ];
    const expected = [
      'Hephaestus VI',
      '40 Eridani A',
      '1,000 km/s',
      'TBD',
      'Launch',
    ];

    it('should display the formatted data', () => {
      const { getAllByRole } = render(
        <DataTableBody columns={columns} data={data} name={name} />,
        { wrapper: BodyWrapper },
      );
      const cells = getAllByRole('cell');

      expect(cells).toHaveLength(5);
      expect(cells.map(cell => cell.textContent)).toEqual(expected);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableBody columns={columns} data={data} name={name} />,
        { wrapper: BodyWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there are many rows', () => {
    const data: SpaceMission[] = [
      {
        deltaV: '1,000 km/s',
        launchDate: '2050',
        missionObjective: 'Titan',
        ordinal: 'IX',
        program: 'Chronos',
      },
      {
        deltaV: '10,000 km/s',
        launchDate: '2150',
        missionObjective: 'Alpha Centauri A',
        ordinal: 'X',
        program: 'Chiron',
      },
      {
        deltaV: '100,000 km/s',
        launchDate: '3050',
        missionObjective: 'Andromeda',
        ordinal: 'XI',
        program: 'Perseus',
      },
    ];
    const expected = [
      'Chronos IX',
      'Titan',
      '1,000 km/s',
      '2050',
      'Launch',
      'Chiron X',
      'Alpha Centauri A',
      '10,000 km/s',
      '2150',
      'Launch',
      'Perseus XI',
      'Andromeda',
      '100,000 km/s',
      '3050',
      'Launch',
    ];

    it('should display the formatted data', () => {
      const { getAllByRole } = render(
        <DataTableBody columns={columns} data={data} name={name} />,
        { wrapper: BodyWrapper },
      );
      const cells = getAllByRole('cell');

      expect(cells).toHaveLength(15);
      expect(cells.map(cell => cell.textContent)).toEqual(expected);
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <DataTableBody columns={columns} data={data} name={name} />,
        { wrapper: BodyWrapper },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with emptyMessage: component', () => {
    const EmptyMessage = (): JSX.Element => (
      <span className="data-table-empty-message">
        You are not going to space today.
      </span>
    );

    describe('when there are no rows', () => {
      const data: SpaceMission[] = [];
      const expected = 'You are not going to space today.';

      it('should display the empty message', () => {
        const { getByRole } = render(
          <DataTableBody
            columns={columns}
            data={data}
            emptyMessage={EmptyMessage}
            name={name}
          />,
          { wrapper: BodyWrapper },
        );
        const cell = getByRole('cell');

        expect(cell.textContent).toBe(expected);
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableBody
            columns={columns}
            data={data}
            emptyMessage={EmptyMessage}
            name={name}
          />,
          { wrapper: BodyWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });

  describe('with emptyMessage: string', () => {
    const emptyMessage = 'You are not going to space today.';

    describe('when there are no rows', () => {
      const data: SpaceMission[] = [];

      it('should display the empty message', () => {
        const { getByRole } = render(
          <DataTableBody
            columns={columns}
            data={data}
            emptyMessage={emptyMessage}
            name={name}
          />,
          { wrapper: BodyWrapper },
        );
        const cell = getByRole('cell');

        expect(cell.textContent).toBe(emptyMessage);
      });

      it('should match the snapshot', () => {
        const { asFragment } = render(
          <DataTableBody
            columns={columns}
            data={data}
            emptyMessage={emptyMessage}
            name={name}
          />,
          { wrapper: BodyWrapper },
        );

        expect(asFragment()).toMatchSnapshot();
      });
    });
  });
});
