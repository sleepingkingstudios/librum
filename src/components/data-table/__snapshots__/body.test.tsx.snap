// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataTableBody /> when there are many rows should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Chronos IX
        </td>
        <td
          class="data-table-cell-mission-objective"
          data-label="Objective"
        >
          Titan
        </td>
        <td
          class="data-table-cell-delta-v text-right lining-nums tabular-nums"
          data-label="Delta V"
        >
          1,000 km/s
        </td>
        <td
          class="data-table-cell-launch-date text-center"
          data-label="Launch Date"
        >
          2050
        </td>
        <td
          class="data-table-cell-actions"
          data-label="Actions"
        >
          <span
            class="data-table-actions"
          >
            Launch
          </span>
        </td>
      </tr>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Chiron X
        </td>
        <td
          class="data-table-cell-mission-objective"
          data-label="Objective"
        >
          Alpha Centauri A
        </td>
        <td
          class="data-table-cell-delta-v text-right lining-nums tabular-nums"
          data-label="Delta V"
        >
          10,000 km/s
        </td>
        <td
          class="data-table-cell-launch-date text-center"
          data-label="Launch Date"
        >
          2150
        </td>
        <td
          class="data-table-cell-actions"
          data-label="Actions"
        >
          <span
            class="data-table-actions"
          >
            Launch
          </span>
        </td>
      </tr>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Perseus XI
        </td>
        <td
          class="data-table-cell-mission-objective"
          data-label="Objective"
        >
          Andromeda
        </td>
        <td
          class="data-table-cell-delta-v text-right lining-nums tabular-nums"
          data-label="Delta V"
        >
          100,000 km/s
        </td>
        <td
          class="data-table-cell-launch-date text-center"
          data-label="Launch Date"
        >
          3050
        </td>
        <td
          class="data-table-cell-actions"
          data-label="Actions"
        >
          <span
            class="data-table-actions"
          >
            Launch
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableBody /> when there are no rows should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-empty"
          colspan="5"
        >
          There are no matching space missions.
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableBody /> when there is one row should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-cell-mission-name"
          data-label="Mission Name"
        >
          Hephaestus VI
        </td>
        <td
          class="data-table-cell-mission-objective"
          data-label="Objective"
        >
          40 Eridani A
        </td>
        <td
          class="data-table-cell-delta-v text-right lining-nums tabular-nums"
          data-label="Delta V"
        >
          1,000 km/s
        </td>
        <td
          class="data-table-cell-launch-date text-center"
          data-label="Launch Date"
        >
          TBD
        </td>
        <td
          class="data-table-cell-actions"
          data-label="Actions"
        >
          <span
            class="data-table-actions"
          >
            Launch
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableBody /> with emptyMessage: component when there are no rows should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-empty"
          colspan="5"
        >
          <span
            class="data-table-empty-message"
          >
            You are not going to space today.
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`<DataTableBody /> with emptyMessage: string when there are no rows should match the snapshot 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="data-table-empty"
          colspan="5"
        >
          You are not going to space today.
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
